### PR TITLE
Alter content negotiation algorithm

### PR DESCRIPTION
**Pull request by tuespetre**
- Always respect explicit Accept headers (except for when they do contain the 'browser wildcard' of `*/*` unless `RespectBrowserAcceptHeader` is configured to `true`, as was the case already), and return `406` in the response when a formatter cannot be found that respects the header
- Remove `HttpNotAcceptableOutputFormatter` and its tests since it appeared to only exist to serve the old algorithm, plus it wasn't really a formatter
- Remove the `FailedContentNegotiation` property from `OutputFormatterCanWriteContext` since it was only used by `HttpNotAcceptableOutputFormatter`
- Remove test cases that confirmed the behavior of using whatever available formatter even if it does not align with an explicit Accept header
- Restructure/consolidate various conneg tests into a single test that receives a series of `ConnegScenario` objects via `[Theory][MemberData]`

### Other considerations

If it is 'too late in the game' to change the behavior this way, or if the current behavior needs to stick around for legacy reasons, it would now be very easy to add a flag along the lines of `RespectAcceptHeaderStrict` with a default of `false`. Then, while selecting a formatter, this flag can be checked at 'the last minute' and used to determine whether or not to grab the first available formatter without any regard to the Accept header.

### CODE REVIEW

**Comment by Eilon** [comment]
@dougbu / @rynowak thoughts on this behavior?

**Comment by rynowak** [comment]
The original design goal here was to make conneg a simple 'system' where as policy as possible was put into formatters rather than the 'system`. That's the line of thinking at gets us the `NoContentFormatter`, the `StringFormatter` and the `NotAcceptableFormatter`. The idea is that these formatters determine the behavior and that the intended way of configuring the system is to add and remove formatters.

Personally, I think the fact that we have `FailedContentNegotiation` and `IgnoreBrowserAcceptHeader` shows that this isn't super tenable. I like what removing `FailedContentNegotiation` does to the design - we can easily make the behavior of whether or not to a do a 'failed' conneg pass an option (as much as I hate boolean switches on options).

Whether or not this is the right default behavior, I want to get more input, but I think this moves the design to a better place.

**Comment by Eilon** [comment]
I chatted with @rynowak and @danroth27 about this.

We feel that if we were to do something for v1.0.0 around this that the change would be more like this:
* Remove `HttpNotAcceptableOutputFormatter`, because no one likes it
* Add a new `bool` option called something like `FallbackToHttp406IfNoFormatterMatches` and have it be `false` by default

The reasons for this are that we think the current behavior is the most frequently requested behavior: you get JSON by default. It is also fully compliant with HTTP because all HTTP says is that the response's content must match the declared response `Content-Type` header. There is no rule that the response has to respect the request's `Accept` headers, or alternatively fail if it can't handle any of them. It's all a "best effort," which we already have.

@tuespetre any thoughts on this? Would you consider updating the PR with this behavior?

**Comment by tuespetre** [comment]
@eilon true, the RFC (at least the 'good old' one) says 'should', not 'must'. I will update this PR with the switch as soon as I get the chance.

**Comment by dougbu** [comment]
Much appreciated @tuespetre

**Comment by tuespetre** [comment]
:up: :date:

@eilon I went the 'opposite' way with the switch and called it `UseFallbackFormatterWhenContentNegotiationFails` with a default of `true`, because it (in my opinion) better reflects the actual behavior of the component. I hope this is acceptable.

**Comment by dougbu** [comment]
@tuespetre we tend to have `bool`s that default to `false`. That avoids having to set the default at all and makes it easier to reason about the code.

**Comment by Eilon** [comment]
I'm not super concerned about whether it's `true` or `false` by default, but the current name one way or another doesn't lend itself well to being a Boolean value. Though the name `FallbackToHttp406IfNoFormatterMatches` was a bit tongue-in-cheek, it made it relatively clear what happens and when if you set it (though perhaps it isn't super clear what happens if it's `false`).

The name `UseFallbackFormatterWhenContentNegotiationFails` doesn't make it very clear to me either way what happens. Plus, it isn't clear what "fails" means in this context: content negotiation couldn't negotiate? Or it threw an exception?

**Comment by tuespetre** [comment]
@eilon yeah this seems to be a prime example of 'naming is hard.' Everything that might describe it well enough seems really long and unwieldy.

**5 minutes later**

Actually, I may have found an applicable short one. We have `RespectBrowserAcceptHeader`, so how about `RespectUnsatisfiableAcceptHeader`? The default would be `false`.

**Comment by dougbu** [comment]
@tuespetre suggest you go ahead w/ `RespectUnsatisfiableAcceptHeader` since @Eilon hasn't objected :smiley_cat:

**Comment by Eilon** [comment]
Sorry, I don't quite get what `RespectUnsatisfiableAcceptHeader` is trying to convey.

**Comment by tuespetre** [comment]
What if we "flip it" so to speak: "DisregardUnsatisfiableAcceptHeader"

"If we cannot satisfy the Accept header, we will disregard the Accept header."

**Comment by Eilon** [comment]
Thanks for everyone's patience here.

I chatted with @danroth27 and we think something like this could work:

```c#
/// Gets or sets a value that decides whether an HTTP 406 Not Acceptable response
/// will be returned if no formatter has been selected to format the response.
public ReturnHttpNotAcceptable { get; set; } // false by default
```

**Comment by tuespetre** [comment]
:up::date:

thanks @Eilon

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs [hunk 1]
```diff
@@ -175,71 +182,57 @@ public virtual Task ExecuteAsync(ActionContext context, ObjectResult result)
                 throw new ArgumentNullException(nameof(formatters));
             }
 
-            // Check if any content-type was explicitly set (for example, via ProducesAttribute
-            // or URL path extension mapping). If yes, then ignore content-negotiation and use this content-type.
-            if (contentTypes.Count == 1)
+            var request = formatterContext.HttpContext.Request;
+            var acceptableMediaTypes = GetAcceptableMediaTypes(contentTypes, request);
+            IOutputFormatter selectedFormatter = null;
+            bool selectFormatterWithoutRegardingAcceptHeader = false;
```
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs]: Van use some more `var` here 😄

**Comment by Eilon** [comment]
@tuespetre thanks!

I'll let @dougbu do the final review on this.

**Comment by dougbu** [comment]
Thanks @Eilon and @tuespetre. @tuespetre I'll likely review this in detail sometime this evening.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs [hunk 2]
```diff
@@ -65,6 +66,11 @@ public class ObjectResultExecutor
         protected bool RespectBrowserAcceptHeader { get; }
 
         /// <summary>
+        /// Gets the value of <see cref="MvcOptions.ReturnHttpNotAcceptable"/>.
+        /// </summary>
+        protected bool ReturnHttpNotAcceptable { get; }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs]: Not for this PR but it's odd that `RespectBrowserAcceptHeader` and now `ReturnHttpNotAcceptable` are `protected` properties when both could be `private` fields. There's no user scenario for subclassing this `.Internal`-namespace type. @tuespetre you've been working w/ this code more recently that me. Am I missing anything?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs [hunk 3]
```diff
@@ -131,6 +137,7 @@ public virtual Task ExecuteAsync(ActionContext context, ObjectResult result)
                 // No formatter supports this.
                 Logger.NoFormatter(formatterContext);
 
+                formatterContext.ContentType = default(StringSegment);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs]: Why is this necessary? Won't be using `formatterContext` when returning a `406` status.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ObjectResultExecutorTest.cs [hunk 4]
```diff
@@ -20,37 +20,6 @@ namespace Microsoft.AspNetCore.Mvc.Internal
 {
     public class ObjectResultExecutorTest
     {
-        [Fact]
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ObjectResultExecutorTest.cs]: I'm having trouble understanding changes to existing tests due to this refactoring. Please restore the old approach so that changes forced by the `src` changes are obvious.

Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ObjectResultExecutorTest.cs]: No problem with `ConnegScenario` in general but would prefer to see it introduced w/o `src` changes in the same PR. (Well, "conneg" is a fairly ugly shorthand but whatever.)

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ObjectResultExecutorTest.cs [hunk 5]
```diff
@@ -75,177 +43,110 @@ public void SelectFormatter_WithNoProvidedContentType_DoesConneg()
...
+                    },
+                    // Respects Accept header value by failing to negotiate an unsupported media type
+                    new ConnegScenario
+                    {
+                        ContentTypes = { "text/plain", "text/xml", "application/json"},
+                        AcceptHeaderValue = "application/vnd.custom",
+                        ResultingContentType = default(StringSegment),
+                        Formatters = { cannotWriteFormatter, jsonFormatter },
+                        ResultingFormatter = null,
+                        ReturnHttpNotAcceptable = true,
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ObjectResultExecutorTest.cs]: Should check the `Response.StatusCode` as well in all of these scenarios -- generally not set but `406` in this scenario.

Should also include `ReturnHttpNotAcceptable = true` in additional variants of the above cases. That'll confirm this setting has no impact when an acceptable formatter is found.

**Comment by dougbu** [comment]
:watch: for a few questions and test changes. Liking the improved `ObjectResultExecutor` readability.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs (hunk 2 above)
Comment (tuespetre) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs]: I'm not sure that they need to be `protected`; I just made the new property `protected` to match what was there :smile_cat:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs (hunk 1 above)
Comment (tuespetre) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs]: Noted; how does the team feel about using `default` to facilitate `var` for `null` values?

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ObjectResultExecutorTest.cs (hunk 5 above)
Comment (tuespetre) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/ObjectResultExecutorTest.cs]: The status code is set by `ExecuteAsync` rather than `SelectFormatter`, and it is already verified by the `ExecuteAsync_NoFormatterFound_Returns406` test.

I will add explicit values for `ReturnHttpNotAcceptable` to verify that the settings does not negatively impact cases where there is an acceptable formatter.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs]: > how does the team feel about using `default` to facilitate `var` for `null` values?

You won't see much use of `default` in MVC at all. Put another way, the `selectedFormatter` declaration is fine.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs (hunk 3 above)
Comment (tuespetre) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs]: This was for testing but since I'm dialing those changes back I will remove it.

**Comment by tuespetre** [comment]
:up::date:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs [hunk 6]
```diff
@@ -175,71 +181,57 @@ public virtual Task ExecuteAsync(ActionContext context, ObjectResult result)
                 throw new ArgumentNullException(nameof(formatters));
             }
 
-            // Check if any content-type was explicitly set (for example, via ProducesAttribute
-            // or URL path extension mapping). If yes, then ignore content-negotiation and use this content-type.
-            if (contentTypes.Count == 1)
+            var request = formatterContext.HttpContext.Request;
+            var acceptableMediaTypes = GetAcceptableMediaTypes(contentTypes, request);
+            var selectedFormatter = default(IOutputFormatter);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectResultExecutor.cs]: FYI only: The main downside of this is some developers read `default` and think the type is a `struct`. In this case, they then have to _think_ about `struct`s and `interface`s :smiling_imp: 

In future I suggest sticking w/ `IOutputFormatter selectedFormatter = null;`, an approach that also works when the variable doesn't need to be initialized.

**Comment by dougbu** [comment]
> I will add explicit values for `ReturnHttpNotAcceptable` to verify that the settings does not negatively impact cases where there is an acceptable formatter.

Where did you do this? I can't see it.

**Comment by tuespetre** [comment]
> Where did you do this? I can't see it.

I stashed away the whole `ConnegScenario` thing and left the test suite with as little modification as possible due to your feedback:

> I'm having trouble understanding changes to existing tests due to this refactoring. Please restore the old approach so that changes forced by the `src` changes are obvious.
>
> No problem with `ConnegScenario` in general but would prefer to see it introduced w/o `src` changes in the same PR. (Well, "conneg" is a fairly ugly shorthand but whatever.)

Let me know if it's cool to add those changes as their own commit to this same pull request and I will do it.

**Comment by dougbu** [comment]
> Let me know if it's cool to add those changes as their own commit to this same pull request and I will do it.

Let's get all of the tests working as-is. Right now a couple of functional tests are failing because `NotAcceptable` is returned where the test expects `Ok`. Should have mentioned this earlier, sorry.

I suggest running `build --quiet verify` on the command line.

Are you saying you need to add `ConnegScenario` in order to add a few tests? I was thinking more in terms of changing a few `[Fact]`s into `[Theory]`s with a `bool returnHttpNotAcceptable` parameter that's used in the options for those tests and shouldn't affect their assertions.

**Comment by tuespetre** [comment]
:up::date: 

Sorry about that, I overlooked that when I was 'squashing' things earlier. The failing functional test was actually removed in the initial PR -- there's not really a good, immediate way to recreate what it did with the removal of the `HttpNotAcceptableOutputFormatter`. I've axed it again. Also tweaked the `default(IOutputFormatter)` bit.

**Comment by dougbu** [comment]
:shipit: we're set and I'll get this in soon.

If we need additional assertions or tests, we can add them later. Same for refactoring the tests. Feel free to make suggestions in the form of test-only PRs.

Much appreciated.

**Comment by dougbu** [comment]
Oops, missed fact the commit description doesn't mention #4612. Please add that reference and, if you have time, make sure your commit is up-to-date w/ `dev`.

Again, much appreciated.

**Comment by tuespetre** [comment]
Both of those are done :smile_cat:

**Comment by Eilon** [comment]
@tuespetre we'll get this on today, thanks!!

**Comment by dougbu** [comment]
14f68b6
